flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/data/StructRowData.java

```diff
@@ -0,0 +1,340 @@
+  }
+
+  @Override
+  public double getDouble(int pos) {
+    return struct.get(pos, Double.class);
+  }
+
+  @Override
+  public StringData getString(int pos) {
+    return isNullAt(pos) ? null : getStringDataInternal(pos);
+  }
+
+  private StringData getStringDataInternal(int pos) {
+    CharSequence seq = struct.get(pos, CharSequence.class);
+    return StringData.fromString(seq.toString());
+  }
+
+  @Override
+  public DecimalData getDecimal(int pos, int precision, int scale) {
+    return isNullAt(pos)
+        ? null
+        : DecimalData.fromBigDecimal(getDecimalInternal(pos), precision, scale);
+  }
+
+  private BigDecimal getDecimalInternal(int pos) {
+    return struct.get(pos, BigDecimal.class);
+  }
+
+  @Override
+  public TimestampData getTimestamp(int pos, int precision) {
+    long timeLong = getLong(pos);
+    if (precision == 6) {
+      int nanosOfMillisecond = (int) (timeLong % 1000);
+      return TimestampData.fromEpochMillis(timeLong / 1000, nanosOfMillisecond);
+    } else {
+      return TimestampData.fromEpochMillis(timeLong);
+    }
+  }
+
+  @Override
+  public <T> RawValueData<T> getRawValue(int pos) {
+    return RawValueData.fromBytes(getBinary(pos));
+  }
+
+  @Override
+  public byte[] getBinary(int pos) {
+    return isNullAt(pos) ? null : getBinaryInternal(pos);
+  }
+
+  private byte[] getBinaryInternal(int pos) {
+    Object bytes = struct.get(pos, Object.class);
+
+    // should only be either ByteBuffer or byte[]
+    if (bytes instanceof ByteBuffer) {
+      return ByteBuffers.toByteArray((ByteBuffer) bytes);
+    } else if (bytes instanceof byte[]) {
+      return (byte[]) bytes;
+    } else {
+      throw new IllegalStateException(
+          "Unknown type for binary field. Type name: " + bytes.getClass().getName());
+    }
+  }
+
+  @Override
+  public ArrayData getArray(int pos) {
+    return isNullAt(pos) ? null : getArrayInternal(pos);
+  }
+
+  private ArrayData getArrayInternal(int pos) {
+    return collectionToArrayData(
+        type.fields().get(pos).type().asListType().elementType(),
+        struct.get(pos, Collection.class));
+  }
+
+  @Override
+  public MapData getMap(int pos) {
+    return isNullAt(pos) ? null : getMapInternal(pos);
+  }
+
+  private MapData getMapInternal(int pos) {
+    return mapToMapData(type.fields().get(pos).type().asMapType(), struct.get(pos, Map.class));
+  }
+
+  @Override
+  public RowData getRow(int pos, int numFields) {
+    return isNullAt(pos) ? null : getStructRowData(pos, numFields);
+  }
+
+  private StructRowData getStructRowData(int ordinal, int numFields) {
+    return new StructRowData(
+        type.fields().get(ordinal).type().asStructType(), struct.get(ordinal, StructLike.class));
+  }
+
+  private MapData mapToMapData(Types.MapType mapType, Map<?, ?> map) {
+    // make a defensive copy to ensure entries do not change
+    List<Map.Entry<?, ?>> entries = ImmutableList.copyOf(map.entrySet());
+
+    ArrayData keyArray =
```

Comment:
why do we need the complex logic from converting map to two lists (key and value) and convert back to a map? what's the problem of simply returning `new GenericMapData(map)`?

Reply:
We need to convert the values in map according to the type of Flink, and put type and value into mapdata. The difference can be seen below:
![image](https://user-images.githubusercontent.com/59213263/205661819-874742df-a9b8-48ae-ad57-2d27e019632a.png)
only `new GenericMapData(map)`

![image](https://user-images.githubusercontent.com/59213263/205661843-846f9be9-6b3a-49cc-8187-fd2c7b442e45.png)
 converting map to two lists

If we just use `new GenericMapData(map)`, when try to run `SELECT summary['flink.job-id'] FROM %s$snapshots`, we will get `ClassCastException: java.lang.String cannot be cast to org.apache.flink.table.data.StringData`

Reply:
ok. I got it that we need run the conversion recursively (e.g. for StringData key or RowData value). I like to propose a diff structure for this class.

We can repurpose `collectionToArrayData` to a `convertFromStructToRowData(Type type, Object value)` method. If `type` is most primitive, it should be simple pass-thru. for other types (like timestamp, struct, map, array) there are some conversions to Flink types (like TimestampData, RowData, MapData, ArrayData). 

Then for the map, we don't have to convert to key and value array. I believe the reason you did it is to reuse the `collectionToArrayData`.

Reply:
Done, please take a look. Thx.